src/services/formatting/formatting.ts

```diff
@@ -439,20 +439,21 @@ namespace ts.formatting {
         }
 
         if (previousRange! && formattingScanner.getStartPos() >= originalRange.end) {
-            const token =
+            const tokenInfo =
                 formattingScanner.isOnEOF() ? formattingScanner.readEOFTokenRange() :
                 formattingScanner.isOnToken() ? formattingScanner.readTokenInfo(enclosingNode).token :
                 undefined;
 
-            if (token) {
+            if (tokenInfo) {
+                const parent = findPrecedingToken(tokenInfo.end, sourceFile, enclosingNode) || previousParent!;
                 processPair(
-                    token,
-                    sourceFile.getLineAndCharacterOfPosition(token.pos).line,
-                    enclosingNode,
+                    tokenInfo,
+                    sourceFile.getLineAndCharacterOfPosition(tokenInfo.pos).line,
+                    parent,
                     previousRange,
                     previousRangeStartLine!,
                     previousParent!,
-                    enclosingNode,
+                    parent,

```

Comment:
The TLDR is this final parameter is supposed to be the “context node” of the token, as if anyone knows what that means, and `enclosingNode` may be arbitrarily higher up the tree. Turns out that “context node” means “direct parent” in almost all cases. What was happening in this case was that `contextNode` should have been the inner `if` Block (`{}`), but it was actually the whole outer `IfStatement` (`if (foo) { ... }`). The rule `NewLineBeforeCloseBraceInBlockContext` is supposed to take effect only if the block ending in the close brace is not on a single line. The inner one is, so the rule should not have applied, but it was using the outer one to decide whether to apply the rule.